simulations.py

```diff
@@ -0,0 +1,247 @@
+
+class Simulation():
+  """An object for a single GPS Simulation.
+
+  Attributes:
+    run_time: simulation duration in seconds
+    gain: signal gain for the broadcast by bladeRF board
+    process: python subprocess created for the simulation
+    process_args: arguments for command to pass into subprocess to run the
+    simulation in bladeGPS
+    start_time: DateTime object for most recent start time
+    end_time: DateTime object for most recent end time
+  """
+
+  def __init__(self, run_time, gain=-2):
+    self.run_time = run_time
+    self.gain = gain
+    self.process = None
+    self.process_args = ["./run_bladerfGPS.sh", "-T", "now", "-d", str(self.run_time)]
+    self.start_time = None
+    self.end_time = None
+
+  def run_simulation(self):
+    """Starts bladeGPS subprocess using given simulation process arguments.
+    """
+    self.start_time = datetime.datetime.utcnow()
+    self.process = subprocess.Popen(self.process_args, stdin=subprocess.PIPE, cwd="./bladeGPS")
+    return
+
+  def end_simulation(self):
+    """Ends the bladeGPS subprocess.
+
+    Checks if the process is still running, and then attempts to quit (via
+    keyboard signal), terminate, and then kill the subprocess in that order
+    """
+    while self.process.poll() is None:
+      self.process.communicate(input="q".encode())
+      print("Quitting simulation...")
+      if self.process.poll() is None:
+        print("Terminating subprocess...")
+        self.process.terminate()
+        time.sleep(2)
+      if self.process.poll() is None:
+        print("Killing subprocess...")
+        self.process.kill()
+        time.sleep(2)
+    self.end_time = datetime.datetime.utcnow()
+    self.process = None
+    print("Subprocess closed.")
+    print("------------------------------------------------")
+
+  def is_running(self):
+    """Checks if there is bladeGPS subprocess currently running.
+
+    Returns:
+      True if there is a running process, false otherwise
+    """
+    return self.process is not None and self.process.poll() is None
+
+
+class StaticSimulation(Simulation):
+  """An object for a single GPS Simulation for a static location.
+
+  Attributes:
+    run_time: simulation duration in seconds
+    gain: signal gain for the broadcast by bladeRF board
+    process: python subprocess created for the simulation
+    process_args: arguments for command to pass into subprocess to run the
+    simulation in bladeGPS
+    start_time: DateTime object for most recent start time
+    end_time: DateTime object for most recent end time
+    latitude: static location latitude in decimal degrees
+    longitude: static location longitude in decimal degrees
+  """
+
+  def __init__(self, run_time, gain, latitude, longitude):
+    Simulation.__init__(self, run_time, gain)
+    self.latitude = latitude
+    self.longitude = longitude
+    coordinate = "%s,%s" % (latitude, longitude)
+    self.process_args = ["./run_bladerfGPS.sh", "-l", coordinate, "-T", "now",
```

Comment:
I think this might break some encapsulation guidelines since the sub-class now has to know implementation details about the superclass, in this case it has to know that the super class wants self.process_args to be the args to run blade_gps.

Since the commands are relatively similar, one approach you could try here is to add a static method or add a class to get it to run the command through composition.

```python
#  ["./run_bladerfGPS.sh", "-T", "now", "-d", str(self.run_time), "-a", str(self.gain)]
#  ["./run_bladerfGPS.sh", "-T", "now", "-d", str(self.run_time), "-a", str(self.gain), "-l", coordinate ]
#  ["./run_bladerfGPS.sh", "-T", "now", "-d", str(self.run_time), "-a", str(self.gain), "-u", file_path ]
               
class BladeRfGpsCommandRunner():               
  # I'm not sure what arguments could be optional to ./run_bladerfGPS.sh, so I'm
  # just assuming the ones that are currently set in Simulation are mandatory
  def __init__(self,
               run_duration,
               gain,
               script_path='./run_bladerfGPS.sh',
               start_time='now',
               location=None,
               coord_file_path=None):
    self._run_duration = run_duration
    self._gain = gain
    ...
    self._coord_file_path = coord_file_path

  def _build_command(self):
    base_cmd = [...]
    if self._location:
      ...
     
  def run_command(self):
    cmd = self._build_command();
    return subprocess.Process(cmd, ...)
```

The command runner would have to be init'd inside a factory methods or overloaded constructors for both StaticSimulation and DynamicSimulation.

Sorry for typing out a huge block of text and code, I just wanted to convince myself that my comment made sense. Doing this with a static method for this would probably look a lot less like Java/C++.

Reply:
Not sure if I totally understand how to fix this, would a function 
```
def _build_bladeGPS_command():
  command = ["./run_bladerfGPS.sh", "-T", "now"]
  if run_duration:
    command.append("-d")
    command.append(str(self.__run_duration))
  if gain:
    command.append("-a")
    command.append(str(self.__gain))
return command
```
And then having overrrides to append location or dynamic file in the subclasses be ok? Is it a problem for self.__process to be modified in the subclasses when creating the process when I overload run_simulation to call subprocess.popen() with the command for that subclass?

Reply:
I think it'd be better to have non-inherited public function `build_bladeGPS_command` that builds the command so that all the tests for it are kept at the same place. This also make it easier to make changes to it in case the path to the script changes.
The problem with overriding `_build_bladeGPS_command` is that we'd be overriding a private (or protected) method which is kind of weird. Modifying the process object is probably more dangerous than modifying the attributes, since the attributes are only used in one place.

We could make `_build_bladeGPS_command(self)` public or just transform `self.process_args` into a method, which would add flexibility and make it safer. Either way, whether you opt for composition, turning `process_args` into a method or keeping it as an attribute you should document how subclass's should be implemented. I can't find any Python specific guidelines for this, but http://go/ej3e-19 applies to all languages.

FYI, we tend to avoid double underscores and use a single one to indicate it's private: go/pystyle#naming-conventions